libevm/legacy/legacy.go

```diff
@@ -31,7 +35,10 @@ func (c PrecompiledStatefulContract) Upgrade() vm.PrecompiledStatefulContract {
 	return func(env vm.PrecompileEnvironment, input []byte) ([]byte, error) {
 		gas := env.Gas()
 		ret, remainingGas, err := c(env, input, gas)
-		if used := gas - remainingGas; used < gas {
+		if remainingGas > gas {
+			return nil, fmt.Errorf("remaining gas %d exceeds supplied gas %d", remainingGas, gas)
+		}
+		if used := gas - remainingGas; used <= gas {
 			env.UseGas(used)
```

Comment:
Should we care about the returned value `hasEnoughGas`? 🤔 I suppose we do check here, but wanted to triple check

Reply:
Depending on the implementation of env though, right? As in it should, but will it always 🤔

Reply:
It's just a proxy for `vm.Contract.UseGas`, which is fundamental to how the EVM does gas accounting. This feels like a "famous last words" scenario though so lets change to:

```go
if !env.UseGas(gas - remainingGas) {
  return nil, 0, vm.ErrOutOfGas
}
```

Reply:
> This feels like a "famous last words" scenario 

I laughed quite a bit over here.

Regarding error wrapping 👀 😈  Could we wrap the error with some context here? Where is this Upgrade() method called, if anywhere? 🤔